src/adr/0034-distributed-listeners-configuration.md

```diff
@@ -0,0 +1,179 @@
+ - Upon modification of the actual topology an "add" or "remove" event is emitted.
+ - Each running James reacts to these events by instantiating the corresponding listener and starting consuming the 
+ associated queue, or stops consuming the associated queue.
+ 
+If a listener is added but is not in the classpath, an ERROR log is emitted. This can happen during a rolling upgrade,
+which defines a new guice binding for a new mailbox listener. Events will still be emitted (and consumed by other James
+servers) however a local James upgrade will be required to effectively be able to start processing these events. The 
+binding will not need to be redefined.
+
+Propagating changes will thus no longer need server reboot.
+
+This won't be contributed in the short run as it requires significant development effort.
+
+### Rolling upgrade scenari
+
+During a rolling upgrade, the james version is heterogeneous across the cluster, and so might be the mailbox listeners
+required at the Guice level.
+
+**case 1**: James Server version 1 does not require listener A, James server version 2 requires listener A.
+
+Since listener A is registered, James server version 1 can not be rebooted prior being upgraded. (As listener A cannot be instanciated)
+
+**case 2**: James Server version 1 requires listener A, James server version 2 does not requires listener A.
+
+Upgrading to James version 2 means  that listener A is still registered as an additional listener, that needs to be 
+manually unconfigured once the rolling upgrade finished. Whish is acceptable in upgrade instruction. We need to make 
+sure the listeners could still be instanciated (even with empty code) for a transition period.
+
+## Error: a listener can't be registered
+
+If a mailbox listener can't be register, eg if its class is not on the classpath, JAMES startup, according to this
+proposal won't be aborted.
+
+This results of an effort to allow an admin can easily unconfigure that given listener using standard webadmin APIs.
+
+However, this will result in a James server not consuming that given group (remote server possessing the JAR will still 
+correctly process all events). This needs a careful log review to get the admin reacting upon the incident.
+
+This result in an incorrect state locally, but the james cluster as a whole still behaves well, as James servers which
+can still run the listener.
+
+We explored the following alternatives:
+
+ - **Failing** upon James startup, but providing a CLI utility to unconfigure a falty group if needed.
+ - **Failing** upon James startup, but running James in a downgraded state allowing to unconfigure the falty listener.
+ 
+The fact that James is failing forces an admin to react, however 'downgraded' status might lead to automated tools to 
+believe James is fully operational, while it is not the case.
+
+We did not chose these options as they imply more development for limited safety gains.
```

Comment:
It deserves a debate: in the past, I implemented a two-stage startup of the software I was working on. The first stage brings admin APIs up, the second stage is starting other services. It's a really powerful system that, in the long run, makes a lot of situations a lot more easier to deal with.

So, yes, it implies more development but maybe it's a good investment.

Reply:
It is loosely coupled with the current ADR content.

Hence, I wrote:

```
If a listener is configured with a class unknown to James, the start-up fails and James starts in a degraded state 
allowing to unconfigure the faulty listener. This downgraded state will be described in a separate ADR and the link will
be updated here.
```